lib/trino-filesystem/src/main/java/io/trino/filesystem/memory/MemoryFileSystem.java

```diff
@@ -0,0 +1,182 @@
+        String key = toBlobKey(location);
+        return new MemoryInputFile(location, () -> blobs.get(key), OptionalLong.of(length));
+    }
+
+    @Override
+    public TrinoOutputFile newOutputFile(String location)
+    {
+        String key = toBlobKey(location);
+        return new MemoryOutputFile(
+                location,
+                new OutputBlob()
+                {
+                    @Override
+                    public boolean exists()
+                    {
+                        return blobs.containsKey(key);
+                    }
+
+                    @Override
+                    public void createBlob(Slice data)
+                            throws FileAlreadyExistsException
+                    {
+                        if (blobs.putIfAbsent(key, new MemoryBlob(data)) != null) {
+                            throw new FileAlreadyExistsException(location);
+                        }
+                    }
+
+                    @Override
+                    public void overwriteBlob(Slice data)
+                    {
+                        blobs.put(key, new MemoryBlob(data));
+                    }
+                });
+    }
+
+    @Override
+    public void deleteFile(String location)
+            throws IOException
+    {
+        if (blobs.remove(toBlobKey(location)) == null) {
+            throw new NoSuchFileException(location);
+        }
+    }
+
+    @Override
+    public void deleteDirectory(String location)
+            throws IOException
+    {
+        String prefix = toBlobPrefix(location);
+        blobs.keySet().removeIf(path -> path.startsWith(prefix));
+    }
+
+    @Override
+    public void renameFile(String source, String target)
+            throws IOException
+    {
+        String sourceKey = toBlobKey(source);
+        String targetKey = toBlobKey(target);
+
+        // File rename is not atomic and that is ok
+        MemoryBlob sourceData = blobs.get(sourceKey);
+        if (sourceData == null) {
+            throw new IOException("File rename from %s to %s failed: Source does not exist".formatted(source, target));
+        }
+        if (blobs.putIfAbsent(targetKey, sourceData) != null) {
+            throw new IOException("File rename from %s to %s failed: Target already exists".formatted(source, target));
+        }
+        blobs.remove(sourceKey, sourceData);
+    }
+
+    @Override
+    public FileIterator listFiles(String location)
+            throws IOException
+    {
+        String prefix = toBlobPrefix(location);
+        Iterator<FileEntry> iterator = blobs.entrySet().stream()
+                .filter(entry -> entry.getKey().startsWith(prefix))
+                .map(entry -> new FileEntry(
+                        "memory:///" + entry.getKey(),
+                        entry.getValue().data().length(),
+                        entry.getValue().lastModified(),
+                        Optional.empty()))
+                .iterator();
+        return new FileIterator() {
```

Comment:
Should we add a `fromIterator()` method on `FileIterator` and make the `empty()` method use that? Then this would be
```java
return FileIterator.fromIterator(iterator);
```
Not sure if this will be useful to other implementations